src/codelabs/async-await/index.md

```diff
@@ -0,0 +1,346 @@
+explanation you can [find out more about
+Future.delayed]({{site.dart_api}}/Future.delayed.html))
+* `createOrderMessage` calls `getUserOrder` as if it were a synchronous function
+ -- it invokes `getUserOrder` and immediately prints the result. 
+* Since `getUserOrder` waits four seconds before returning the user's order,
+`createOrderMessage` never actually prints the user's order. Instead the console
+prints the value immediately returned by `createOrderMessage` which
+turns out to be `Instance of '_Future<String>'`.
+
+What is a future, and how do you write code to handle the asynchronous function
+ `getUserOrder`?
+ 
+ {{ site.alert.secondary }} 
+ Quick review: 
+ * Asynchronous operations let a program do work while waiting for an operation to finish.
+ * In Dart, you can use Futures, `async` and `await` to write declarative asynchronous code.
+ {{ site.alert.end }}
+
+
+## Async in Dart: What is a future?
+
+In Dart, a
+[Future]({{site.dart_api}}/Future-class.html)
+represents the result of an asynchronous operation. A future object has
+two states: uncompleted or completed.
+
+{{ site.alert.note }}
+'Uncompleted' is a special term in Dart referring to the state of a future 
+ before it has produced a result.
+{{ site.alert.end }}
+
+### Uncompleted
+
+When you invoke a function that returns a future, that function immediately 
+returns a future object in an uncompleted state. While in its uncompleted state,
+a future object is simply waiting for asynchronous operations to finish or to 
+throw an error. 
+
+### Completed
+
+There are two ways that a future object can complete:
+
+#### 1. Success
+
+If the asynchronous operations succeed, the future completes and returns a 
+result of type T. For example,  
+`Future<String>` is the type signature for a future object that produces a string
+result. If a future produces a result that isn't a usable value, then the
+future's type is `Future<void>`. 
+
+#### 2. Error
+
+If the asynchronous operations being performed by the 
+function throw an error, the future completes and returns an error. 
+
+To see this lifecycle at work consider what happens after invoking the 
+`getUserOrder` function:
+
+
+<iframe style="margin-top: 25px; margin-bottom: 25px" frameborder="no" height="450" src="https://dartpad.dartlang.org/experimental/embed-new.html?id=57e6085344cbd1719ed42b32f8ad1bce" width="100%"></iframe>
+
+
+After 1 second has passed, notice that the future is in an uncompleted state - you can't access the
+desired result yet:
+
+`     Your order is: Instance of '_Future<String>'`
+
+After 4 seconds have passed, the future completes and produces the desired
+result:
+
+`Your order is: 'Large Latte'`
+
+How do you access this result that completes 4 seconds later? In the next section
+you will learn about using  the `async` and `await` keywords to handle the results
+of asynchronous functions. 
+
+{{ site.alert.secondary }}
+Quick Review:
```

Comment:
-i  like your quick review; i think I understand what your writing about at a high level. i'm curious, though, about why a Future is called a Future?

Reply:
The pattern has been around for a while: https://en.wikipedia.org/wiki/Futures_and_promises

While I don't think that reviewing the history would be useful in the codelab, perhaps an additional sentence or two explaining how the name relates to the concept might help?